### ISSUE TITLE
Best practice for taking advantage of EF first-level caching in Domain Service Layer.

### ISSUE THREAD

**Issue by crush83** [mehdime/DbContextScope#58]
Part of the appeal of using EF as opposed to a micro-ORM like Dapper is the first-level caching (to my knowledge, Dapper doesn't do this).

However, by placing the DbContextScope (or any Unit of Work style pattern) in my domain service layer methods, acting as the business transaction boundary, I'm not sure how to take advantage of EF's caching.

For example, consider the following interface:

```
public interface IOrderService
{
    Order Get(int orderId);
    void MarkOrderApproved(Order order, User user);
}
```
Imagine I then have a controller action processing a request to mark an order approved. It receives the `orderId` and I have the `userId` from the authenticated user making the call. However, I don't have the entities yet, and my domain service contract wants those entities. Thus:

```
public class OrderController
{
    public IActionResult MarkOrderApproved(int orderId)
    {
        var order = orderService.Get(orderId);
        var user = userService.Get(User.Id); //However this works.

        orderService.MarkOrderApproved(order, user);

        return Ok();
    }
}
```

What's the best practice here? Is it to not write my Domain Service methods to accept entities, and allow each method to do whatever lookups inside the scope? That would allow it take advantage of the first-level cache, and my code becomes:

```
public class OrderController
{
    public IActionResult MarkOrderApproved(int orderId)
    {
        orderService.MarkOrderApproved(orderId, User.Id);
        return Ok();
    }
}
```

and

```
public class OrderService : IOrderService
{
    void MarkOrderApproved(int orderId, int userId)
    {
        using (var dbContextScope = dbContextScopeFactory.Create())
        {
            var order = orderRepository.Get(orderId);
            var user = userRepository.Get(userId);

            order.Status = OrderStatus.Approved;

            dbContextScope.SaveChanges();
        }
    }
}
```
This seems like a clean solution, but it restricts my domain service layer from accepting domain entities as parameters so that they may be looked up within the internal scope.

Should I offer overloads on my domain service interface, like:
```
public interface IOrderService
{
    void MarkOrderApproved(int orderId, int userId);
    void MarkOrderApproved(Order order, User user);
}
```

This seems dirty to me, but I'm curious what the best practice is for taking advantage of the first level caching while not making my domain service layer suffer.

Is there another option that I'm missing outside of method injection of the DbContext (which this library is meant to solve)?

*EDIT*
Some sites seem to suggest that domain services should only accept primitives, enums, structs, and DTOs as arguments to their methods.

DDD seems to suggest that they should only accept Domain Objects, and that ORM entities are DTOs that should be mapped to Domain Objects before being passed to the domain service layer.

To me this indicates that any transactions should occur outside of domain services, at the application service layer exclusively. Domain services would be ignorant of transactions.

Thanks.

**Comment by VictorioBerra** [comment]
Personally, I think in this situation your service should accept the IDs and, perform the lookups and make the changes it needs all in the context of one transaction. After all, all of that is a single business transaction like Mehdi argues.

**Comment by crush83** [comment]
After doing a ton of research on DDD, I came to the conclusion that there are basically two ideologies in the industry today: pure and hybrid DDD.

The pure ideology seems to dictate that the ORM layer should be separated from the domain layer because:

- It leaks persistence details (like database row ids) that aren't necessarily relevant to your domain model.
- It makes your DDD layer dependent on whatever ORM technology you've chosen to use.

One major benefit of following the pure ideology is that you pass fully resolved or constructed domain objects into your domain services. You can be sure that these objects are valid because they've passed all of the domain logic required to create one.

One major drawback is that you have to maintain the mapping from ORM object to Domain object somewhat manually. A library like AutoMapper can help with this mapping, but you still have to do at least somewhat basic configuration.

Depending on your point of view, the separation of ORM from Domain layer can be seen as a benefit or a drawback. If the persistence layer changes, then the domain model isn't immediately impacted. And vice versa. If you are doing a lot of changes with a medium sized team of different knowledge levels, then this could be a nuisance. as programmers forget to update the mapping code. However, it can also save your domain model from getting polluted with persistence fields that aren't relevant to the given object. You can split a single ORM entity into multiple domain objects this way as well (although this complicates the write back to the database)

TL;DR It depends on how tightly coupled with your ORM you are willing to make your Domain layer. I found that ideally, it's best to pass full domain objects to domain services only, but id's are fine in the application services.

Abstracting the DbContextScope into a more general transaction UoW framework can also allow you to pass it down through your domain layer services without coupling them to EF.

**Comment by VictorioBerra** [comment]
@crush83 

That is good findings. Seem's like you spent some time researching the purpose of DDD. You are correct that separation of concerns is a good idea.

**If there is one thing you get out of all this, in the end none of this really matters. The most important thing is making it easier and cleaner to TEST your app.**

I have seen it argued that planning to swap out your persistence layer is silly and will lead to further unnecessary abstractions and mappings. It leads to the repository pattern which I personally dislike and it's the whole reason I have chosen to explore this library. In my mind, this library blurs the line a bit between the repo pattern and the service layer. It lets me have much finer control over my transactions and I get a place to store my queries away from my business logic and it also lets me easily integration test because **I can mock the transactions and rollback on save. I want a decent abstraction, and high test-ability, and I don't want to over-architect my app and add additional complexity because some day I might remove EF and use `<insert newest amazing ORM system>`.**

I do have some comments on your points:
> he pure ideology seems to dictate that the ORM layer should be separated from the domain layer because
> * It leaks persistence details (like database row ids) that aren't necessarily relevant to your domain model.

When using EF, you will always have a domain layer that resembles your persistence layer. That is just how EF works. and that has a tendency to drive all your other layers. Unless you are using something like Mongo, your app will sort of reflect your relational database.

I don't think there is anything wrong with using an ID to manipulate your entity instead of mapping and passing around a bunch of stuff. After all, many of your entities will need a unique identifier to manage their persistence. Why would you abstract that away.

By the way, there is an alternative to n-tier onion style architecture https://msdn.microsoft.com/en-us/library/jj591573.aspx they do warn about the drawbacks to.

**Comment by crush83** [comment]
@VictorioBerra 

I had some arguments with some DDD _experts_ on the requirement of having an immutable identity to every object. They seem to neglect that after you are done manipulating the domain object in your domain layer, you eventually need to write it back to the persistence layer, which requires an immutable identity of some kind to accomplish the relational mapping. If that is lost in abstraction, then you have no way to accomplish that task. It could be argued in this case that the identity becomes a domain object detail as well in my opinion. Whether that immutable identity is an auto-increment integer or some other unique key doesn't really matter.

**I'm really glad you brought up CQRS.** I certainly ran across this in my research, and have some questions/concerns with it. I mostly see it referenced as an alternative to the Repository pattern. Though, it seems that you could use it in conjunction with the Repository pattern as well if that fits your model.

I like the total separation of queries and commands into their own classes in CQRS. I think this makes code more readable, and easier to maintain. It also makes it so only dependencies needed for the given query or command exist on the class instead of injecting dependencies that other queries/commands might need, but your specific command/query does not.

Some questions I need further direction on include:

- Do you write interfaces for every query type, or just a single generic query interface (likewise for commands)? How do you mock these in this case (just mock the generic interface)?
- How do you manage dependencies for your queries/commands?
    - Build some kind of query/command factory that has access to your IoC if you are using one?
- May Commands/Queries call other Commands/Queries? Is there an argument for and against this type of behavior?

Also, I realize this might not be the best forum for continuing this discussion, but not sure where else we could continue. Do you have suggestions? Maybe StackOverflow question or something of that nature?

Thanks for the comments, much appreciated.

**Comment by VictorioBerra** [comment]
I don't know much about the CQRS pattern unfortunately. I think StackOverflow might be you best bet. Maybe a couple Slack channels, maybe #csharp on freenode.net
